logger_test.go

```diff
@@ -95,3 +101,123 @@ func TestLogger_SetBufferPool(t *testing.T) {
 	assert.Equal(t, pool.get, 1, "Logger.SetBufferPool(): The BufferPool.Get() must be called")
 	assert.Len(t, pool.buffers, 1, "Logger.SetBufferPool(): The BufferPool.Put() must be called")
 }
+
+func TestLogger_concurrentLock(t *testing.T) {
+	SetFormatter(&LogFormatter{})
+	go func() {
+		for {
+			func() {
+				defer func() {
+					// 处理所有异常，防止panic导致程序关闭
+					if p := recover(); p != nil {
+					}
+				}()
+				hook := AddTraceIdHook("123")
+				defer RemoveTraceHook(hook)
+				Infof("test why ")
+			}()
+		}
+	}()
+	go func() {
+		for {
+			func() {
+				defer func() {
+					// 处理所有异常，防止panic导致程序关闭
+					if p := recover(); p != nil {
+					}
+				}()
+				hook := AddTraceIdHook("1233")
+				defer RemoveTraceHook(hook)
+				Infof("test why 2")
+			}()
+		}
+	}()
+	time.Sleep(1 * time.Minute)
+}
+
+var traceLock = &sync.Mutex{}
+
+func AddTraceIdHook(traceId string) Hook {
+	defer traceLock.Unlock()
+	traceLock.Lock()
+	traceHook := newTraceIdHook(traceId)
+	if StandardLogger().Hooks == nil {
+		hooks := new(LevelHooks)
+		StandardLogger().ReplaceHooks(*hooks)
+	}
+	AddHook(traceHook)
+	return traceHook
+}
+
+func RemoveTraceHook(hook Hook) {
+	allHooks := StandardLogger().Hooks
+	func() {
+		defer Unlock()
+		Lock()
+		for key, hooks := range allHooks {
+			replaceHooks := hooks
+			for index, h := range hooks {
+				if h == hook {
+					replaceHooks = append(hooks[:index], hooks[index:]...)
+					break
+				}
+			}
+			allHooks[key] = replaceHooks
+		}
+	}()
+
+	StandardLogger().ReplaceHooks(allHooks)
+}
+
+type TraceIdHook struct {
+	TraceId string
+	GID     uint64
+}
+
+func newTraceIdHook(traceId string) Hook {
+	return &TraceIdHook{
+		TraceId: traceId,
+		GID:     getGID(),
+	}
+}
+
+func (t TraceIdHook) Levels() []Level {
+	return AllLevels
+}
+
+func (t TraceIdHook) Fire(entry *Entry) error {
+	if getGID() == t.GID {
+		entry.Context = context.WithValue(context.Background(), "trace_id", t.TraceId)
+	}
+	return nil
+}
+
+type LogFormatter struct{}
+
+//格式详情
+func (s *LogFormatter) Format(entry *Entry) ([]byte, error) {
+	timestamp := time.Now().Format("2006-01-02 15:04:05")
+	var file string
+	var line int
+	if entry.Caller != nil {
+		file = filepath.Base(entry.Caller.File)
+		line = entry.Caller.Line
+	}
+	level := entry.Level.String()
+	if entry.Context == nil || entry.Context.Value("trace_id") == "" {
+		uuid := "NO UUID"
+		entry.Context = context.WithValue(context.Background(), "trace_id", uuid)
+	}
+	msg := fmt.Sprintf("%-15s [%-3d] [%-5s] [%s] %s:%d %s\n", timestamp, getGID(), level, entry.Context.Value("trace_id"), file, line, entry.Message)
+	return []byte(msg), nil
+}
+
+// 获取当前协程id
```

Comment:
and this?

Reply:
I have deleted unnecessary Chinese comments